optuna/_optimize.py

```diff
@@ -193,6 +196,20 @@ def _run_trial(
     func_err_fail_exc_info: Optional[Any] = None
     # Set to a string if `func` returns correctly but the return value violates assumptions.
     values_conversion_failure_message: Optional[str] = None
+    stop_event: Optional[Event] = None
+    thread: Optional[Thread] = None
+
+    if (
+        study._storage.is_heartbeat_supported()
+        and study._storage.get_heartbeat_interval() is not None
+    ):
+        study._storage.record_heartbeat(trial._trial_id)
```

Comment:
Isn't this now always guaranteed to be invoked in the beginning of each thread?

Reply:
I've noticed that we need it because it allows joining the TrialHeartbeatModel with TrialModel for the current running trial as follows.
```python
stale_trials = (
    session.query(models.TrialModel)
    .join(models.TrialModel.heartbeats)
    .filter(models.TrialModel.state == TrialState.RUNNING)
    .filter(func.now() - models.TrialHeartbeatModel.heartbeat > grace_period)
    .all()
)
```

Reply:
Sorry but are we not still calling this method in the beginning of the thread? What is the difference?